File: 977. Squares of a Sorted Array/solution1/solution.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        z = 0
        while z < n and nums[z] < 0:
            z += 1
        res = []
        l, r = z - 1, z
        while l >= 0 and r < n:
            if abs(nums[l]) <= nums[r]:
                res.append(nums[l] * nums[l])
                l -= 1
            else:
                res.append(nums[r] * nums[r])
                r += 1

        while l >= 0:
            res.append(nums[l] * nums[l])
            l -= 1
        while r < n:
            res.append(nums[r] * nums[r])
            r += 1
        return res

File: 977. Squares of a Sorted Array/solution1/Solution.java
import java.util.List;

class Solution {
  public int[] sortedSquares(int[] nums) {
    int n = nums.length;
    int[] res = new int[n];
    // Find the z, `nums[z] == 0`
    int z = 0;
    while (z < n && nums[z] < 0)
      z++;
    int l = z - 1, r = z;
    int i = 0;
    while (l >= 0 && r < n) {
      if (Math.abs(nums[l]) <= Math.abs(nums[r])) {
        res[i++] = nums[l] * nums[l];
        l--;
      } else {
        res[i++] = nums[r] * nums[r];
        r++;
      }
    }
    while (l >= 0) {
      res[i++] = nums[l] * nums[l];
      l--;
    }
    while (r < n) {
      res[i++] = nums[r] * nums[r];
      r++;
    }

    return res;
  }
}